### PR TITLE
[GLUTEN-7362][VL] Add test for 'IN' and 'OR' filter in Scan

### PR DESCRIPTION
**Pull request by zml1206**
## What changes were proposed in this pull request?

Fixes: \#7362
Resolved by #6754, add test for it.

## How was this patch tested?

UT

### CODE REVIEW

**Comment by FelixYBW** [comment]
Is this a Gluten issue?

**Comment by zml1206** [comment]
> Is this a Gluten issue?

Yes, the first singularOrList in `OR` needs to check in range.

**Comment by zml1206** [comment]
cc @rui-mo Thank you.

File: backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala [hunk 1]
```diff
@@ -239,6 +239,13 @@ class MiscOperatorSuite extends VeloxWholeStageTransformerSuite with AdaptiveSpa
       "select l_orderkey from lineitem " +
         "where l_partkey in (1552, 674) or l_partkey in (1552) and l_orderkey > 1") { _ => }
     checkLengthAndPlan(df, 73)
+
+    runQueryAndCompare(
+      "select count(1) from lineitem " +
+        "where (l_shipmode in ('TRUCK', 'MAIL') or l_shipmode in ('AIR', 'FOB')) " +
+        "and l_shipmode in ('RAIL','SHIP')") {
```
Comment (rui-mo) [inline backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala:246]: I think this issue is in the condition `(A or B) and C`, when C is already pushed down, the `(A or B)` cannot be pushed down.

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc [hunk 2]
```diff
@@ -1669,7 +1669,8 @@ bool SubstraitToVeloxPlanConverter::canPushdownOr(
   }
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxPlan.cc:1669]: And I suggest we add a `setOrRange` function in RangeRecorder which returns false when other conditions already exist for a field.

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc [hunk 3]
```diff
@@ -1686,14 +1687,15 @@ bool SubstraitToVeloxPlanConverter::canPushdownOr(
...
     } else if (arg.value().has_singular_or_list()) {
       const auto& singularOrList = arg.value().singular_or_list();
+      // Disable IN pushdown for int-like types.
       if (!canPushdownSingularOrList(singularOrList, true)) {
         return false;
       }
       uint32_t fieldIdx = getColumnIndexFromSingularOrList(singularOrList);
-      // Disable IN pushdown for int-like types.
-      if (!rangeRecorders.at(fieldIdx).setInRange(true /*forOrRelation*/)) {
+      if (!rangeRecorders.at(fieldIdx).setInRange(sign /*forOrRelation*/)) {
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxPlan.cc:1695]: We don't need this change as this logic is checking the nested expressions in an `or` condition, while here the issue is `or` cannot be pushed down totally.

**Review by rui-mo — COMMENTED** [review]
Thanks. These are my thoughts.

File: backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala (hunk 1 above)
Comment (zml1206) [reply backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala:246]: Yes, but only for `in`, which is `singularOrList`.

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc (hunk 2 above)
Comment (zml1206) [reply cpp/velox/substrait/SubstraitToVeloxPlan.cc:1669]: `setInRange` itself retains a parameter `forOrRelation`, which is used for verification. We only need to judge the first singularOrList in or.

Comment (zml1206) [reply cpp/velox/substrait/SubstraitToVeloxPlan.cc:1669]: Thanks for your suggestion, updated.

File: cpp/velox/substrait/SubstraitToVeloxPlan.h [hunk 4]
```diff
@@ -286,6 +286,16 @@ class SubstraitToVeloxPlanConverter {
       return true;
     }
 
+    /// Set the existence of or-range and returns whether it can coexist with
+    /// existing conditions for this field.
+    bool setOrRange() {
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxPlan.h:291]: Sorry but I just notice 'setMultiRange' is for the 'or' condition, so we can just use it.

**Review by rui-mo — COMMENTED** [review]
Thanks.

File: cpp/velox/substrait/SubstraitToVeloxPlan.h (hunk 4 above)
Comment (zml1206) [reply cpp/velox/substrait/SubstraitToVeloxPlan.h:291]: `setMultiRange` seems to conflict with `setCertainRangeForFunction` of 1681L.

File: backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala [hunk 5]
```diff
@@ -239,6 +239,13 @@ class MiscOperatorSuite extends VeloxWholeStageTransformerSuite with AdaptiveSpa
       "select l_orderkey from lineitem " +
         "where l_partkey in (1552, 674) or l_partkey in (1552) and l_orderkey > 1") { _ => }
     checkLengthAndPlan(df, 73)
+
+    runQueryAndCompare(
```
Comment (rui-mo) [inline backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala:243]: We have merged https://github.com/apache/incubator-gluten/commit/d6326f06c8158de2eb08bb9ba7e3e4eb6e3964ef. Would you like to try if this issue has been fixed? Perhaps we can merge this test to ensure the functionality. Thanks for your patience.

Comment (zml1206) [reply backends-velox/src/test/scala/org/apache/gluten/execution/MiscOperatorSuite.scala:243]: Thank you, I will try it locally later.

**Comment by zml1206** [comment]
Local test passed, I modified pr to only add ut. @rui-mo

**Pull request merged**